### PR TITLE
feat: datanode stats is stored in the mem_kv of meta leader

### PR DESCRIPTION
**Pull request by Fengys123**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This pr main change:
1. datanode stats is store in the mem_kv store of the meta leader, instead of etcd kv store.
2. add a cluster grpc service.
3. add an http api to query all heartbeat data(or datanode stats).
```
curl http://127.0.0.1:3002/admin/heartbeat
```

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/meta-srv/src/error.rs [hunk 1]
```diff
@@ -165,6 +165,33 @@ pub enum Error {
         #[snafu(backtrace)]
         source: BoxedError,
     },
+
+    #[snafu(display("Invalid KVs length, expect: {}", expect))]
+    InvalidKvsLength { expect: usize, backtrace: Backtrace },
```
Comment (MichaelScofield) [inline src/meta-srv/src/error.rs:170]: better carry actual kvs length here

File: src/meta-srv/src/error.rs [hunk 2]
```diff
@@ -165,6 +165,33 @@ pub enum Error {
...
+    BatchGet {
+        source: tonic::Status,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display("Response header not found"))]
+    ResponseHeaderNotFound { backtrace: Backtrace },
+
+    #[snafu(display("The requested meta node is not leader"))]
+    IsNotLeader { backtrace: Backtrace },
```
Comment (MichaelScofield) [inline src/meta-srv/src/error.rs:191]: better carry the node's addr

File: src/meta-srv/src/keys.rs [hunk 3]
```diff
@@ -174,7 +174,7 @@ impl<'a> TableRouteKey<'a> {
     }
 }
 
-#[derive(Eq, PartialEq, Debug, Hash)]
+#[derive(Eq, PartialEq, Debug, Clone, Hash)]
```
Comment (MichaelScofield) [inline src/meta-srv/src/keys.rs:177]: better derived `Copy`, so no `clone` is needed

File: src/meta-srv/src/service/admin/heartbeat.rs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            .as_ref()
+            .context(error::NoMetaPeerClientSnafu)?;
+        let stat_kvs = meta_peer_client.get_all_dn_stat_kvs().await?;
+        let stat_vals: Vec<StatValue> = stat_kvs.into_values().collect();
+        let result = StatValues { stat_vals }.try_into()?;
+
+        Ok(http::Response::builder()
+            .status(http::StatusCode::OK)
+            .body(result)
+            .unwrap())
```
Comment (MichaelScofield) [inline src/meta-srv/src/service/admin/heartbeat.rs:44]: Is `unwrap` safe here?

File: src/meta-srv/src/service/cluster.rs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+impl MetaSrv {
+    pub fn is_leader(&self) -> bool {
+        if let Some(election) = self.election() {
+            if election.is_leader() {
+                return true;
+            }
+        }
+        false
```
Comment (MichaelScofield) [inline src/meta-srv/src/service/cluster.rs:72]:
```suggestion
        self.election().map(|x| x.is_leader()).unwrap_or(false)
```

File: src/meta-srv/src/metasrv.rs [hunk 6]
```diff
@@ -100,20 +101,23 @@ pub struct MetaSrv {
     selector: SelectorRef,
     handler_group: HeartbeatHandlerGroup,
     election: Option<ElectionRef>,
+    meta_peer_client: Option<MetaPeerClient>,
 }
 
 impl MetaSrv {
     pub async fn new(
```
Comment (MichaelScofield) [inline src/meta-srv/src/metasrv.rs:108]: suggest using a builder pattern here; args are too long

File: src/meta-srv/src/keys.rs (hunk 3 above)
Comment (Fengys123) [reply src/meta-srv/src/keys.rs:177]: The definition of the `Copy` trait is as follows: 
```
pub trait Copy: Clone
```
If implement `Copy`, must implement `Clone`

File: src/api/greptime/v1/meta/cluster.proto [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+// Cluster service is used for communication between meta nodes.
+service Cluster {
+  // Batch get kvs by input keys from leader's in_memory kv store. 
+  rpc BatchGet(GetKvRequest) returns (GetKvResponse);
+
+  // Range get the kvs from leader's in_memory kv store.
+  rpc Range(RangeRequest) returns (RangeResponse);
+}
+
+message GetKvRequest {
```
Comment (fengjiachun) [inline src/api/greptime/v1/meta/cluster.proto:17]: `BatchGetRequest` ?

File: src/api/greptime/v1/meta/cluster.proto [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  RequestHeader header = 1;
+
+  repeated bytes keys = 2;
+}
+
+message GetKvResponse {
```
Comment (fengjiachun) [inline src/api/greptime/v1/meta/cluster.proto:23]: `BatchGetResponse` ?

File: src/meta-srv/src/metasrv/builder.rs [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    CheckLeaderHandler, CollectStatsHandler, HeartbeatHandlerGroup, KeepLeaseHandler,
+    OnLeaderStartHandler, PersistStatsHandler, ResponseHeaderHandler,
+};
+use crate::metasrv::{ElectionRef, MetaSrv, MetaSrvOptions, SelectorRef, TABLE_ID_SEQ};
+use crate::selector::lease_based::LeaseBasedSelector;
+use crate::sequence::Sequence;
+use crate::service::store::kv::{KvStoreRef, ResetableKvStoreRef};
+use crate::service::store::memory::MemStore;
+
+#[allow(dead_code)]
```
Comment (MichaelScofield) [inline src/meta-srv/src/metasrv/builder.rs:29]: Seems `MetaSrvBuilder` is already been used.

```suggestion
```

**Review by MichaelScofield — APPROVED** [review]

**Review by fengjiachun — APPROVED** [review]

**Pull request merged**